src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -140,6 +148,31 @@ void GcsServer::Start() {
   gcs_init_data->AsyncLoad([this, gcs_init_data] { DoStart(*gcs_init_data); });
 }
 
+void GcsServer::CacheAndSetClusterId() {
+  static std::string const kTokenNamespace = "cluster";
+  kv_manager_->GetInstance().Get(
+      kTokenNamespace, kClusterIdKey, [this](std::optional<std::string> token) mutable {
+        if (!token.has_value()) {
+          ClusterID cluster_id = ClusterID::FromRandom();
+          RAY_LOG(INFO) << "No existing server token found. Generating new token: "
+                        << cluster_id.Hex();
+          kv_manager_->GetInstance().Put(kTokenNamespace,
+                                         kClusterIdKey,
+                                         cluster_id.Binary(),
+                                         false,
+                                         [this, &cluster_id](bool added_entry) mutable {
+                                           RAY_CHECK(added_entry)
+                                               << "Failed to persist new token!";
+                                           rpc_server_.SetClusterId(cluster_id);
```

Comment:
How do we ensure that the rpc server doesn't handle any requests before the cluster id is set?

Reply:
Now that future is removed, we just crash if we expect a cluster ID and don't have one when the service is registered (`grpc_server.cc`, RegisterService). Enforcement-by-crash instead of enforcement-by-hanging.